packages/backend/src/services/PromoteService/PromoteService.ts

```diff
@@ -0,0 +1,761 @@
+                upstreamChart,
+            );
+            return updatedChart;
+        } catch (e) {
+            await this.trackAnalytics(
+                user,
+                'promote.error',
+                promotedChart,
+                upstreamChart,
+                e.message,
+            );
+            throw e;
+        }
+    }
+
+    private async upsertDashboard(
+        user: SessionUser,
+        promotedContent: PromotedDashboard,
+        upstreamContent: UpstreamDashboard,
+    ) {
+        const promotedDashboard = promotedContent.dashboard;
+        if (upstreamContent.dashboard === undefined) {
+            // There are no dashboards with the same slug, we create the chart, and the space if needed
+            // We only check the org/project access
+
+            const space = await this.getOrCreateSpace(
+                user,
+                promotedContent,
+                upstreamContent,
+            );
+            // Create new dashboard
+            const newDashboardData: CreateDashboard & {
+                slug: string;
+            } = {
+                ...promotedContent.dashboard,
+                spaceUuid: space.uuid,
+                slug: promotedContent.dashboard.slug,
+            };
+            const newDashboard = await this.dashboardModel.create(
+                space.uuid,
+                newDashboardData,
+                user,
+                upstreamContent.projectUuid,
+            );
+
+            return newDashboard;
+        }
+
+        // We override existing dashboard details
+        const upstreamDashboard = upstreamContent.dashboard;
+
+        if (
+            upstreamDashboard.name !== promotedDashboard.name ||
+            upstreamDashboard.description !== promotedDashboard.description ||
+            upstreamDashboard.spaceUuid !== upstreamContent.space?.uuid
+        ) {
+            // We also update dashboard name and description if they have changed
+            await this.dashboardModel.update(upstreamDashboard.uuid, {
+                name: promotedDashboard.name,
+                description: promotedDashboard.description,
+                spaceUuid: upstreamContent.space?.uuid,
+            });
+        }
+
+        const updatedChart = await this.dashboardModel.addVersion(
+            upstreamDashboard.uuid,
+            promotedDashboard,
+            user,
+            upstreamContent.projectUuid,
+        );
+
+        return updatedChart;
+    }
+
+    async promoteDashboard(user: SessionUser, dashboardUuid: string) {
+        const dashboard = await this.dashboardModel.getById(dashboardUuid);
+
+        const { upstreamProjectUuid } = await this.projectModel.getSummary(
+            dashboard.projectUuid,
+        );
+        if (!upstreamProjectUuid)
+            throw new NotFoundError(
+                'This chart does not have an upstream project',
+            );
+
+        const promotedSpace = await this.spaceModel.getSpaceSummary(
```

Comment:
we could extract most of the code in this method into a method like `getPromoteCharts` , to make it more readable